blockchain/message_pool/src/msgpool.rs

```diff
@@ -0,0 +1,1066 @@
+    pub async fn load_local(&mut self) -> Result<(), Error> {
+        for (key, value) in self.local_msgs.clone() {
+            self.add(&value).await.unwrap_or_else(|err| {
+                if err == Error::SequenceTooLow {
+                    warn!("error adding message: {:?}", err);
+                    self.local_msgs.remove(&key);
+                }
+            });
+        }
+        Ok(())
+    }
+}
+
+/// Remove a message from pending given the from address and sequence
+pub async fn remove(
+    from: &Address,
+    pending: Arc<RwLock<HashMap<Address, MsgSet>>>,
+    sequence: u64,
+) -> Result<(), Error> {
+    let mut pending = pending.write().await;
+
+    let mset = pending
+        .get_mut(from)
+        .ok_or_else(|| Error::InvalidFromAddr)?;
+    mset.msgs.remove(&sequence);
+
+    if mset.msgs.is_empty() {
+        pending.remove(from);
+    } else {
+        let mut max_sequence: u64 = 0;
+        for sequence_temp in mset.msgs.keys().cloned() {
+            if max_sequence < sequence_temp {
+                max_sequence = sequence_temp;
+            }
+        }
+        if max_sequence < sequence {
+            max_sequence = sequence;
+        }
+        mset.next_sequence = max_sequence + 1;
+    }
+    Ok(())
+}
+
+/// Attempt to get a signed message that corresponds to an unsigned message in bls_sig_cache
+async fn recover_sig(
+    bls_sig_cache: Arc<RwLock<LruCache<Cid, Signature>>>,
+    msg: UnsignedMessage,
+) -> Result<SignedMessage, Error> {
+    let mut bls_sig_cache = bls_sig_cache.write().await;
+    let val = bls_sig_cache
+        .get(&msg.cid()?)
+        .ok_or_else(|| Error::Other("Could not recover sig".to_owned()))?;
+    Ok(SignedMessage::new_from_parts(msg, val.clone()))
+}
+
+/// Finish verifying signed message before adding it to the pending mset hashmap. If an entry
+/// in the hashmap does not yet exist, create a new mset that will correspond to the from message
+/// and push it to the pending hashmap
+async fn add_helper<T>(
+    api: Arc<RwLock<T>>,
+    bls_sig_cache: Arc<RwLock<LruCache<Cid, Signature>>>,
+    pending: Arc<RwLock<HashMap<Address, MsgSet>>>,
+    msg: SignedMessage,
+) -> Result<(), Error>
+where
+    T: Provider,
+{
+    let mut pending = pending.write().await;
+    let api = api.read().await;
+    let mut bls_sig_cache = bls_sig_cache.write().await;
+    if msg.signature().signature_type() == SignatureType::BLS {
+        bls_sig_cache.put(msg.cid()?, msg.signature().clone());
+    }
+
+    if msg.message().gas_limit() > 100_000_000 {
+        return Err(Error::Other(
+            "given message has too high of a gas limit".to_string(),
+        ));
+    }
+
+    api.put_message(&msg)?;
+    let msett = pending.get_mut(msg.message().from());
+    match msett {
+        Some(mset) => mset.add(&msg)?,
+        None => {
+            let mut mset = MsgSet::new();
+            mset.add(&msg)?;
+            pending.insert(*msg.message().from(), mset);
+        }
+    }
```

Comment:
```suggestion
    if msg.signature().signature_type() == SignatureType::BLS {
        bls_sig_cache
            .write()
            .await
            .put(msg.cid()?, msg.signature().clone());
    }

    if msg.message().gas_limit() > 100_000_000 {
        return Err(Error::Other(
            "given message has too high of a gas limit".to_string(),
        ));
    }

    api.read().await.put_message(&msg)?;
    match pending.write().await.get_mut(msg.message().from()) {
        Some(mset) => mset.add(&msg)?,
        None => {
            let mut mset = MsgSet::new();
            mset.add(&msg)?;
            pending.write().await.insert(*msg.message().from(), mset);
        }
    }
```
Is there a reason you have the locks setup as is and not something like this? should really think carefully about how you are using all of these locks, a lot of changes in this PR seem like it's asking for a deadlock

Reply:
I cleaned up the locks a bit, the match statement can't be like you proposed because you get a deadlock because the write lock in 550 is still held if the function is in the none case and needs to run line 555